02_knight_and_princess/robert_gaisin/main.cpp

```diff
@@ -0,0 +1,196 @@
+#include <ctime>
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+using VertexId = int;
+using EdgeId = int;
+
+std::ostream& operator<<(std::ostream& out, const vector<int>& int_vector) {
+  for (auto it = int_vector.begin(); it != int_vector.end(); ++it) {
+    out << *it;
+    if (it != int_vector.end() - 1) {
+      out << ", ";
+    }
+  }
+  return out;
+}
+
+struct Vertex {
+  const VertexId id;
+  int depth = 0;
+  vector<EdgeId> edge_ids;
+};
+
+std::ostream& operator<<(std::ostream& out, const Vertex& vertex) {
+  out << "{" << endl;
+  out << "      \"id\": " << vertex.id << "," << endl;
+  out << "      \"edge_ids\": [" << vertex.edge_ids << "]" << endl;
+  out << "    }";
+  return out;
+}
+
+struct Edge {
+  const EdgeId id;
+  VertexId begin = 0;
+  VertexId end = 0;
+};
+
+std::ostream& operator<<(std::ostream& out, const Edge& edge) {
+  out << "{" << endl;
+  out << "      \"id\": " << edge.id << "," << endl;
+  out << "      \"vertex_ids\": [" << edge.begin << ", " << edge.end << "]"
+      << endl;
+  out << "    }";
+  return out;
+}
+
+class Graph {
+ public:
+  void add_vertex();
+
+  void add_edge(VertexId begin, VertexId end);
```

Comment:
Да, фундаментальные типы лучше предавать по значению, но так как мы спрятали их за кастомным типом, который может измениться, лучше не предполагать, а просто передавать их по `const` ссылке, как это обычно делается для всех не фундаментальных типов.
```suggestion
  void add_edge(const VertexId& begin, const VertexId& end);
```